test/es-module/test-esm-dynamic-import.js

```diff
@@ -0,0 +1,96 @@
+// Flags: --experimental-modules --harmony-dynamic-import
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const { URL } = require('url');
+const vm = require('vm');
+
+common.crashOnUnhandledRejection();
+
+const relativePath = './test-esm-ok.mjs';
+const absolutePath = require.resolve('./test-esm-ok.mjs');
+const targetURL = new URL('file:///');
+targetURL.pathname = absolutePath;
+
+function expectErrorProperty(result, propertyKey, value) {
+  Promise.resolve(result)
+    .catch(common.mustCall(error => {
+      assert.equal(error[propertyKey], value);
+    }));
+}
+
+function expectMissingModuleError(result) {
+  expectErrorProperty(result, 'code', 'ERR_MISSING_MODULE');
+}
+
+function expectInvalidUrlError(result) {
+  expectErrorProperty(result, 'code', 'ERR_INVALID_URL');
+}
+
+function expectInvalidProtocolError(result) {
+  expectErrorProperty(result, 'code', 'ERR_INVALID_PROTOCOL');
+}
+
+function expectInvalidContextError(result) {
+  expectErrorProperty(result, 'message', 'invalid_context');
+}
+
+function expectOkNamespace(result) {
+  Promise.resolve(result)
+    .then(common.mustCall(ns => {
+      // Can't deepStrictEqual because ns isn't a normal object
+      assert.deepEqual(ns, { default: true });
+    }));
+}
+
+function expectFsNamespace(result) {
+  Promise.resolve(result)
+    .then(common.mustCall(ns => {
+      assert.equal(typeof ns.default.writeFile, 'function');
+    }));
+}
+
+expectOkNamespace(import(relativePath));
+
+expectFsNamespace(import("fs"));
+
+expectInvalidProtocolError(import("node:fs"));
+
+// Fails, no valid referrer (indirect eval) & relative path
+expectMissingModuleError(eval.call(null, `import("${relativePath}")`));
+
+// Fails, no valid referrer (direct eval) & relative path
+expectMissingModuleError(eval(`import("${relativePath}")`));
+
+// Succeeds, doesn't need a valid referrer
+expectFsNamespace(eval.call(null, 'import("fs")'));
+
+// Fails because referrer is undefined (and loader/search.js rejects it)
+// TODO(jkrems): This doesn't seem like expected behavior
+expectMissingModuleError(eval.call(null, `import("${targetURL}")`));
+
+// Succeeds
+expectFsNamespace(vm.runInThisContext('import("fs")'));
+
+// Succeeds because it's got an valid base url
+expectOkNamespace(vm.runInThisContext(`import("${relativePath}")`, {
+  filename: __filename,
+}));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
```

Comment:
It looks like `require` handle cases like this without a base url (uses the cwd()?)
Would that be an option here?
```js
expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
```

Reply:
@jdalton 

> It looks like `require` handle cases like this without a base url

Can you provide a test case where `require` works?

Reply:
```shell
$ touch foo.js
$ node
vm.runInThisContext(`require('./foo.js')`)
```

Reply:
@jdalton Ah, that `require` is actually created by the REPL as a global function with the CWD as base. This function is not available in CJS modules executed by `node file.js`.

Reply:
Out of the repl `touch bar.js` then
```js
//bar.js
vm = require("vm")
req = require
console.log(vm.runInThisContext(`req('./foo.js')`))
```
Then `node bar.js`.

Reply:
Node's `require` works because of `path.resolve` use which will use `cwd()`.

Reply:
@jdalton That has nothing to do with the CWD, but rather with where `bar.js` is... which in your case is the CWD.

Reply:
@TimothyGu Cool! It is of its context though and does work. Would dynamic import be able to cue off of `import.meta.url` to get its bearings similar to `require()`?

Reply:
`import.meta.url` is a writable property in [web spec](https://html.spec.whatwg.org/#hostgetimportmetaproperties), you should use a private bound variable instead if you are an environment.

Reply:
> `import.meta.url` is a writable property in web spec, you should use a private bound variable instead if you are an environment.

Yep yep!

Reply:
@jdalton Just to make sure I understand the `bar.js` example correctly: You are assigning the `require` from the module closure to a global `req` and then use that to require something. I'm not sure I follow how this relates to `import`. `import` cannot be assigned or forwarded. The only context we have is the eval/script context of the code running `import`. So I don't think it would make sense to make relative imports work where the importing script pretty explicitly doesn't have a location.

Reply:
@jkrems IMO if it's doable in cjs it would be rad to be doable in esm. I'm for consistency though, so if that means deprecating some aspect in cjs that's better than leaving something doable in cjs but more cumbersome in esm.

Reply:
You can do the same thing you showed above in esm:

```js
//bar.js
vm = require("vm")
req = specifier => import(specifier);
vm.runInThisContext(`req('./foo.js')`).then(console.log)
```

The above should just work.

Reply:
@jkrems Nice! On a separate note, any thoughts on import() with absolute file urls erroring? That still seems like an odd one to error on.

Reply:
Well, it *would* be possible to make it do something but it would mean introducing a completely new mode ("resolve but without an actual base") and I traced it through a handful of layers where things might need to change. And it's a feature that requires disabling all checks along the way for something that *normally* shouldn't happen.

The shorter version: I agree that it's something that's reasonable to expect and I'd be happy to put it on the list. My proposal right now would be "let it throw in the initial PR, carefully make it work in a follow-up".

Reply:
@jkrems 
>The shorter version: I agree that it's something that's reasonable to expect and I'd be happy to put it on the list. My proposal right now would be "let it throw in the initial PR, carefully make it work in a follow-up".

I can dig that!